src/bio/bio_buffer.c

```diff
@@ -1982,3 +1983,43 @@ bio_copy(struct bio_io_context *ioctxt, struct umem_instance *umem,
 
 	return rc;
 }
+
+#define IO_MONITOR_INTVL   1000000  /* us, 1 second */
+#define IO_TIMEOUT_DEFAULT 40000000 /* us, 40 seconds */
+
+void
+bio_io_monitor(struct bio_xs_context *xs_ctxt, uint64_t now)
+{
+	enum smd_dev_type        st;
+	struct bio_xs_blobstore *bxb;
+	struct bio_io_lug       *io_lug;
+	struct media_error_msg  *mem;
+
+	if ((xs_ctxt->bxc_io_monitor_ts + IO_MONITOR_INTVL) > now)
+		return;
+
+	xs_ctxt->bxc_io_monitor_ts = now;
+
+	for (st = SMD_DEV_TYPE_DATA; st < SMD_DEV_TYPE_MAX; st++) {
+		bxb = xs_ctxt->bxc_xs_blobstores[st];
+
+		if (!bxb || d_list_empty(&bxb->bxb_pending_ios))
+			continue;
+
+		io_lug = d_list_entry(bxb->bxb_pending_ios.next, struct bio_io_lug, bil_link);
```

Comment:
Could this access of list entry safe? accessing happen in the blobstore owner xstream, but adding and deleting entry from list might happen on other xstreams. Looks not safe for me?

Reply:
we try to track every xstream IO, but stalling IO checking only happen on owner xstream.

Reply:
I missed that currently monitoring happen on each xstream, thanks for explain.